File: src/minigames/fine_tuning/src/main/java/work/lclpnet/ap2/game/fine_tuning/FineTuningRoom.java
package work.lclpnet.ap2.game.fine_tuning;

import com.mojang.math.Transformation;
import lombok.Getter;
import lombok.Setter;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Vec3i;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Display;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.properties.NoteBlockInstrument;
import org.jetbrains.annotations.Nullable;
import org.joml.Matrix4f;
import work.lclpnet.ap2.game.fine_tuning.melody.FakeNoteBlockPlayer;
import work.lclpnet.ap2.game.fine_tuning.melody.Melody;
import work.lclpnet.ap2.game.fine_tuning.melody.Note;
import work.lclpnet.ap2.impl.util.ColorUtil;
import work.lclpnet.gaco.dynamic_entities.DynamicEntityManager;
import work.lclpnet.gaco.dynamic_entities.PlayerSpecificDynamicEntity;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;

import static java.lang.Math.abs;
import static java.lang.Math.max;

class FineTuningRoom {

    @Getter
    private final BlockPos pos;
    private final BlockPos spawn;
    private final float yaw;
    private final List<@Nullable PlayerSpecificDynamicEntity<Display.BlockDisplay>> displays = new ArrayList<>(5);
    private BlockPos[] noteBlocks = null;
    private int[] notes = null, tmpNotes = null;
    private NoteBlockInstrument[] instruments = null;
    private FakeNoteBlockPlayer nbPlayer = null;
    private boolean temporary = false;
    @Setter @Getter
    private @Nullable BlockPos testSignPos = null;

    public FineTuningRoom(BlockPos pos, BlockPos spawn, float yaw) {
        this.pos = pos;
        this.spawn = spawn;
        this.yaw = yaw;
    }

    public void setNoteBlocks(Vec3i[] relNoteBlock) {
        BlockPos[] noteBlocks = new BlockPos[relNoteBlock.length];

        for (int j = 0; j < relNoteBlock.length; j++) {
            noteBlocks[j] = pos.offset(relNoteBlock[j]);
        }

        this.noteBlocks = noteBlocks;

        this.notes = new int[noteBlocks.length];
        Arrays.fill(notes, 0);

        this.tmpNotes = new int[noteBlocks.length];
        Arrays.fill(tmpNotes, 0);

        instruments = new NoteBlockInstrument[noteBlocks.length];
        Arrays.fill(instruments, NoteBlockInstrument.HARP);

        this.nbPlayer = new FakeNoteBlockPlayer(noteBlocks, notes, instruments);

        displays.clear();

        for (int i = 0; i < notes.length; i++) {
            displays.add(null);
        }
    }

    public void teleport(ServerPlayer player, ServerLevel world) {
        double x = spawn.getX() + 0.5, y = spawn.getY(), z = spawn.getZ() + 0.5;

        player.teleportTo(world, x, y, z, Set.of(), yaw, 0.0F, true);
    }

    public void useNoteBlock(ServerPlayer player, BlockPos pos, DynamicEntityManager manager) {
        int index = getNoteBlock(pos);

        if (index == -1) return;

        int transpose = player.isShiftKeyDown() ? -1 : 1;

        setNote(index, notes[index] + transpose);

        playNote(player, index);
        removeDisplay(index, manager);
    }

    public void playNoteBlock(ServerPlayer player, BlockPos pos) {
        int index = getNoteBlock(pos);

        if (index == -1) return;

        playNote(player, index);
    }

    public void playNote(ServerPlayer player, int index) {
        nbPlayer.play(player, index);
    }

    public void setNote(int i, int note) {
        notes[i] = Math.floorMod(note, 25);
    }

    public int getNoteBlock(BlockPos pos) {
        int index = -1;

        for (int i = 0, noteBlocksLength = noteBlocks.length; i < noteBlocksLength; i++) {
            BlockPos noteBlock = noteBlocks[i];

            if (noteBlock.equals(pos)) {
                index = i;
                break;
            }
        }

        return index;
    }

    public void setTemporaryMelody(Melody melody) {
        System.arraycopy(notes, 0, tmpNotes, 0, notes.length);
        temporary = true;

        setMelody(melody);
    }

    public void restoreMelody() {
        if (!temporary) return;

        System.arraycopy(tmpNotes, 0, notes, 0, tmpNotes.length);
        temporary = false;
    }

    public void setMelody(Melody melody) {
        nbPlayer.setMelody(melody);
    }

    public Melody getCurrentMelody() {
        restoreMelody();

        NoteBlockInstrument instrument = instruments[0];
        Note[] currentNotes = new Note[notes.length];

        var allNotes = Note.values();

        for (int i = 0; i < currentNotes.length; i++) {
            currentNotes[i] = allNotes[notes[i]];
        }

        return new Melody(instrument, currentNotes);
    }

    public int calculateScore(Melody baseMelody, Melody reference) {
        restoreMelody();

        Note[] refNotes = reference.notes();
        int score = 0;

        for (int i = 0; i < notes.length; i++) {
            int actual = notes[i];
            int expected = refNotes[i].ordinal();
            int base = baseMelody.notes()[i].ordinal();

            int offset = abs(expected - base);
            int diff = abs(expected - actual);

            int noteScore = max(0, offset - diff);

            score += noteScore;
        }

        return score;
    }

    public boolean isComplete(Melody reference) {
        restoreMelody();

        Note[] refNotes = reference.notes();

        for (int i = 0; i < notes.length; i++) {
            int actual = notes[i];
            int expected = refNotes[i].ordinal();

            if (actual != expected) {
                return false;
            }
        }

        return true;
    }

    public void markErrors(Melody baseMelody, Melody reference, DynamicEntityManager manager, ServerPlayer player) {
        removeDisplays(manager);

        restoreMelody();

        Note[] refNotes = reference.notes();

        for (int i = 0; i < notes.length; i++) {
            int actual = notes[i];
            int expected = refNotes[i].ordinal();
            int base = baseMelody.notes()[i].ordinal();

            int offset = abs(expected - base);
            int diff = abs(expected - actual);

            if (diff == 0 || offset == 0) continue;

            float error = (float) diff / offset;
            addDisplay(i, error, manager, player);
        }
    }

    public void addDisplay(int note, float error, DynamicEntityManager manager, ServerPlayer viewer) {
        if (note < 0 || note >= displays.size()) return;

        BlockPos pos = noteBlocks[note];
        float margin = 0.015f;

        var display = new Display.BlockDisplay(EntityType.BLOCK_DISPLAY, viewer.level());
        display.setPosRaw(pos.getX() + margin, pos.getY() + margin, pos.getZ() + margin);
        display.setBlockState(Blocks.NOTE_BLOCK.defaultBlockState());
        display.setTransformation(new Transformation(new Matrix4f().scale(1 - margin * 2)));
        display.setGlowingTag(true);

        int color = ColorUtil.lerpRgb(0xefe409, 0x890404, error);
        display.setGlowColorOverride(color);

        var dynamic = new PlayerSpecificDynamicEntity<>(display, viewer.getUUID());

        displays.set(note, dynamic);
        manager.add(dynamic);
    }

    public void removeDisplay(int note, DynamicEntityManager manager) {
        if (note < 0 || note >= displays.size()) return;

        var display = displays.get(note);

        if (display == null) return;

        displays.set(note, null);
        manager.remove(display);
    }

    public void removeDisplays(DynamicEntityManager manager) {
        for (int i = 0; i < displays.size(); i++) {
            var display = displays.get(i);

            if (display == null) continue;

            manager.remove(display);
            displays.set(i, null);
        }
    }
}

File: src/minigames/fine_tuning/src/main/kotlin/work/lclpnet/ap2/game/fine_tuning/FineTuningRoom.kt
package work.lclpnet.ap2.game.fine_tuning

import com.mojang.math.Transformation
import net.minecraft.core.BlockPos
import net.minecraft.core.Vec3i
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Display
import net.minecraft.world.entity.EntityTypes
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.state.properties.NoteBlockInstrument
import org.joml.Matrix4f
import work.lclpnet.ap2.game.fine_tuning.melody.FakeNoteBlockPlayer
import work.lclpnet.ap2.game.fine_tuning.melody.Melody
import work.lclpnet.ap2.game.fine_tuning.melody.Note
import work.lclpnet.ap2.impl.util.ColorUtil
import work.lclpnet.gaco.dynamic_entities.DynamicEntityManager
import work.lclpnet.gaco.dynamic_entities.PlayerSpecificDynamicEntity
import kotlin.math.abs
import kotlin.math.max

class FineTuningRoom(val pos: BlockPos, private val spawn: BlockPos, private val yaw: Float) {

    private val displays = ArrayList<PlayerSpecificDynamicEntity<Display.BlockDisplay>?>(5)
    private lateinit var noteBlocks: Array<BlockPos>
    private lateinit var notes: IntArray
    private lateinit var tmpNotes: IntArray
    private lateinit var instruments: Array<NoteBlockInstrument>
    private lateinit var nbPlayer: FakeNoteBlockPlayer
    private var temporary = false
    var testSignPos: BlockPos? = null

    fun setNoteBlocks(relNoteBlocks: Array<out Vec3i>) {
        noteBlocks = Array(relNoteBlocks.size) { j -> pos.offset(relNoteBlocks[j]) }
        notes = IntArray(noteBlocks.size)
        tmpNotes = IntArray(noteBlocks.size)
        instruments = Array(noteBlocks.size) { NoteBlockInstrument.HARP }
        nbPlayer = FakeNoteBlockPlayer(noteBlocks, notes, instruments)

        displays.clear()
        repeat(notes.size) { displays.add(null) }
    }

    fun teleport(player: ServerPlayer, world: ServerLevel) {
        player.teleportTo(world, spawn.x + 0.5, spawn.y.toDouble(), spawn.z + 0.5, emptySet(), yaw, 0f, true)
    }

    fun useNoteBlock(player: ServerPlayer, pos: BlockPos, manager: DynamicEntityManager): Boolean {
        val index = getNoteBlock(pos)
        if (index == -1) return false

        val transpose = if (player.isShiftKeyDown) -1 else 1
        setNote(index, notes[index] + transpose)
        playNote(player, index)
        removeDisplay(index, manager)
        return true
    }

    fun playNoteBlock(player: ServerPlayer, pos: BlockPos): Boolean {
        val index = getNoteBlock(pos)
        if (index == -1) return false
        playNote(player, index)
        return true
    }

    fun playNote(player: ServerPlayer, index: Int) {
        nbPlayer.play(player, index)
    }

    fun setNote(i: Int, note: Int) {
        notes[i] = Math.floorMod(note, 25)
    }

    fun getNoteBlock(pos: BlockPos): Int {
        for (i in noteBlocks.indices) {
            if (noteBlocks[i] == pos) return i
        }
        return -1
    }

    fun setTemporaryMelody(melody: Melody) {
        System.arraycopy(notes, 0, tmpNotes, 0, notes.size)
        temporary = true
        setMelody(melody)
    }

    fun restoreMelody() {
        if (!temporary) return
        System.arraycopy(tmpNotes, 0, notes, 0, tmpNotes.size)
        temporary = false
    }

    fun setMelody(melody: Melody) {
        nbPlayer.setMelody(melody)
    }

    fun getCurrentMelody(): Melody {
        restoreMelody()
        val instrument = instruments[0]
        val allNotes = Note.entries.toTypedArray()
        val currentNotes = Array(notes.size) { i -> allNotes[notes[i]] }
        return Melody(instrument, currentNotes)
    }

    fun calculateScore(baseMelody: Melody, reference: Melody): Int {
        restoreMelody()
        val refNotes = reference.notes
        var score = 0

        for (i in notes.indices) {
            val actual = notes[i]
            val expected = refNotes[i].ordinal
            val base = baseMelody.notes[i].ordinal
            val offset = abs(expected - base)
            val diff = abs(expected - actual)
            score += max(0, offset - diff)
        }

        return score
    }

    fun correctNoteCount(reference: Melody): Int {
        restoreMelody()
        var correct = 0

        for (i in notes.indices) {
            val actual = notes[i]
            val expected = reference.notes[i].ordinal

            if (actual == expected) {
                correct++
            }
        }

        return correct
    }

    fun isComplete(reference: Melody): Boolean {
        restoreMelody()
        val refNotes = reference.notes

        for (i in notes.indices) {
            if (notes[i] != refNotes[i].ordinal) return false
        }

        return true
    }

    fun markErrors(baseMelody: Melody, reference: Melody, manager: DynamicEntityManager, player: ServerPlayer) {
        removeDisplays(manager)
        restoreMelody()

        val refNotes = reference.notes

        for (i in notes.indices) {
            val actual = notes[i]
            val expected = refNotes[i].ordinal
            val base = baseMelody.notes[i].ordinal
            val offset = abs(expected - base)
            val diff = abs(expected - actual)

            if (diff == 0 || offset == 0) continue

            val error = diff.toFloat() / offset
            addDisplay(i, error, manager, player)
        }
    }

    fun addDisplay(note: Int, error: Float, manager: DynamicEntityManager, viewer: ServerPlayer) {
        if (note < 0 || note >= displays.size) return

        val pos = noteBlocks[note]
        val margin = 0.015f

        val display = Display.BlockDisplay(EntityTypes.BLOCK_DISPLAY, viewer.level())
        display.setPosRaw(pos.x + margin.toDouble(), pos.y + margin.toDouble(), pos.z + margin.toDouble())
        display.blockState = Blocks.NOTE_BLOCK.defaultBlockState()
        display.setTransformation(Transformation(Matrix4f().scale(1 - margin * 2)))
        display.setGlowingTag(true)

        val color = ColorUtil.lerpRgb(0xefe409, 0x890404, error)
        display.glowColorOverride = color

        val dynamic = PlayerSpecificDynamicEntity(display, viewer.uuid)
        displays[note] = dynamic
        manager.add(dynamic)
    }

    fun removeDisplay(note: Int, manager: DynamicEntityManager) {
        if (note < 0 || note >= displays.size) return
        val display = displays[note] ?: return
        displays[note] = null
        manager.remove(display)
    }

    fun removeDisplays(manager: DynamicEntityManager) {
        for (i in displays.indices) {
            val display = displays[i] ?: continue
            manager.remove(display)
            displays[i] = null
        }
    }
}
